# Phylogenomic analysis: From raw data to time-calibrated tree

This is my notebook describing the process of building a phylogeny using target capture and published whole-genome resequencing data. I will use a number of software and packages written in _Python_, _Bash_, _Perl_. The MetaCentrum Czech national cluster provides the computational resources and infrastructure for all the analyses.

## The data

The primary dataset comes from our sampling and target capture approach. The baits to capture come from Espeland et al. (2018), the so-called Anchored Hybrid Enrichment (AHE) BUTTERFLY 1.0 kit., which targeted 425 loci.

As a complementary dataset, Emmanuel Toussaint shared their sequences of several Eudaminae (Lepidoptera: Hesperiidae) taxa. Their data come from a target capture approach using the BUTTERFLY 1.0 kit, and another kit targeting a subset of the 425 loci, the so-called legacy butterfly gene markers.

The published genomes come from Li et al. (2019), which used a whole-genome resequencing approach in 250 representatives of the skipper butterflies (Hesperiidae).

## The MetaCentrum



### References

- Espeland et al. (2018) A comprehensive and dated phylogenomic analysis of butterflies. Current Biology 28(5): 770&ndash;778. DOI: [10.1016/j.cub.2018.01.061](https://doi.org/10.1016/j.cub.2018.01.061)

- Li et al. (2019) Genomes of skipper butterflies reveal extensive convergence of wing patterns. PNAS 116 (13) 6232&ndash;6237. DOI: [10.1073/pnas.1821304116](https://doi.org/10.1073/pnas.1821304116)
