# Installations
You will find the steps to follow all the pipelines, packages, and software needed in our protocol.

### Manipulation of public NCBI data

- [SRA toolkit](https://github.com/pavelm14/Eudaminae_phylogeny/blob/master/bioinformatics/installations/SRAtoolkit.md) to download and manipulate SRA data.

### Pipeline raw data to alignment

- [SECAPR](https://github.com/pavelm14/Eudaminae_phylogeny/blob/master/bioinformatics/installations/SECAPR.md) (including Miniconda2, with Python 2.7)

### Manipulation of Fasta files

- [SeqKit](https://github.com/pavelm14/Eudaminae_phylogeny/blob/master/bioinformatics/installations/SeqKit.md) to edit fasta files in batch (e.g., renaming taxa in every file).

### Phylogeny inference

- [ASTRAL](https://github.com/pavelm14/Eudaminae_phylogeny/blob/master/bioinformatics/installations/ASTRAL.md) to infer species trees from gene trees using a summary approach of the multispecies coalescent.

- [IQTREE 2](https://github.com/pavelm14/Eudaminae_phylogeny/blob/master/bioinformatics/installations/IQTREE.md) to infer phylogenies using the concatenation method.
